ibis/mapd/client.py

```diff
@@ -0,0 +1,326 @@
+
+
+class MapDClient(SQLClient):
+    """
+
+    """
+    database_class = Database
+    sync_query = MapDQuery
+    dialect = MapDDialect
+
+    def __init__(
+        self, uri: str=None, user: str=None, password: str=None,
+        host: str=None, port: int=9091, dbname: str=None,
+        protocol: str='binary', execution_type: int=3
+    ):
+        """
+
+        :param uri:
+        :param user:
+        :param password:
+        :param host:
+        :param port:
+        :param dbname:
+        :param protocol:
+        :param execution_type:
+        """
+        self.uri = uri
+        self.user = user
+        self.password = password
+        self.host = host
+        self.port = port
+        self.dbname = dbname
+        self.protocol = protocol
+        self.execution_type = execution_type
+
+        self.con = pymapd.connect(
+            uri=uri, user=user, password=password, host=host,
+            port=port, dbname=dbname, protocol=protocol
+        )
+
+    @property
+    def _table_expr_klass(self):
+        return ir.TableExpr
+
+    def log(self, msg):
+        log(msg)
+
+    def close(self):
+        """Close MapD connection and drop any temporary objects"""
+        self.con.close()
+
+    def _build_ast(self, expr, context):
+        """
+        Required.
+
+        :param expr:
+        :param context:
+        :return:
+        """
+        result = build_ast(expr, context)
+        return result
+
+    def _fully_qualified_name(self, name, database):
+        if bool(fully_qualified_re.search(name)):
+            return name
+
+        database = database or self.current_database
+        return '{0}.{1}'.format(database, name)
+
+    def _get_table_schema(self, table_name):
+        """
+
+        :param table_name:
+        :return:
+        """
+        database = None
+        table_name_ = table_name.split('.')
+        if len(table_name_) == 2:
+            database, table_name = table_name_
+        return self.get_schema(table_name, database)
+
+    def _execute(self, query, results=False):
+        """
+
+        :param query:
+        :return:
+        """
+        if not self.execution_type == 3:
+            raise NotImplemented()
+
+        return MapDCursor(self.con.cursor().execute(query))
+
+    def database(self, name=None):
+        """Connect to a database called `name`.
+
+        Parameters
+        ----------
+        name : str, optional
+            The name of the database to connect to. If ``None``, return
+            the database named ``self.current_database``.
+
+        Returns
+        -------
+        db : Database
+            An :class:`ibis.client.Database` instance.
+
+        Notes
+        -----
+        This creates a new connection if `name` is both not ``None`` and not
+        equal to the current database.
+        """
+        if name == self.current_database or (
+            name is None and name != self.current_database
```

Comment:
Can `self.current_database` be `None`? If not, then you don't need the second part of the `and` here since `name is None` would imply that `name != self.current_database`.